packages/vite/src/node/server/__tests__/watcher.spec.ts

```diff
@@ -61,25 +34,35 @@ describe('watcher configuration', () => {
   })
 
   it('should watch the root directory, config file dependencies, dotenv files, and the public directory', async () => {
-    await createServer({
-      server: {
-        watch: {},
-      },
-      publicDir: '__test_public__',
-    })
-    expect(watchSpy).toHaveBeenLastCalledWith(
-      expect.arrayContaining(
-        [
-          process.cwd(),
-          resolve('fake/config/dependency.js'),
-          resolve('.env'),
-          resolve('.env.local'),
-          resolve('.env.development'),
-          resolve('.env.development.local'),
-          resolve('__test_public__'),
-        ].map((file) => file.replace(/\\/g, '/')),
-      ),
-      expect.anything(),
+    const root = fileURLToPath(
+      new URL('./fixtures/watcher/nested-root', import.meta.url),
     )
+    server = await createServer({ root })
+    await new Promise((resolve) => server!.watcher.once('ready', resolve))
+    // At this point, there's still a chance that chokidar has not watch all the necessary directories yet
+    // so we have to retry here for a bit
+    await withRetry(() => {
```

Comment:
Vitest has `vi.waitFor` by the way. Not a stopper

This particular case doesn't depend on CI/non-CI (so limiting to 3 retries doesn't make a lot of sense to me), this is just how `ready` event works in chokidar (in module runner test we actually wait for `getWatched` to include a file we are testing without any timeouts)

Reply:
I knew Vitest has a solution for this, but I couldn't find it. I'll try to refactor to use it then. Not very happy with the `withRetry` solution here either